smac_planner/include/smac_planner/costmap_downsampler.hpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2020, Carlos Luis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#ifndef SMAC_PLANNER__DOWNSAMPLER_HPP_
+#define SMAC_PLANNER__DOWNSAMPLER_HPP_
+
+#include <algorithm>
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+#include "smac_planner/constants.hpp"
+
+namespace smac_planner
+{
+
+/**
+ * @class smac_planner::CostmapDownsampler
+ * @brief A costmap downsampler for more efficient path planning
+ */
+class CostmapDownsampler
+{
+public:
+  /**
+   * @brief A constructor for CostmapDownsampler
+   */
+  explicit CostmapDownsampler()
+  : _costmap(nullptr),
+    _downsampled_costmap(nullptr),
+    _downsampled_costmap_pub(nullptr)
+  {
+  }
+
+  /**
+   * @brief A destructor for CostmapDownsampler
+   */
+  ~CostmapDownsampler()
+  {
+  }
+
+  /**
+   * @brief Initialize the downsampled costmap object and the ROS publisher
+   * @param node Lifecycle node pointer
+   * @param global_frame The ID of the global frame used by the costmap
+   * @param topic_name The name of the topic to publish the downsampled costmap
+   * @param costmap The costmap we want to downsample
+   * @param downsampling_factor Multiplier for the costmap sresolution
+   */
+   void initialize(
+    const nav2_util::LifecycleNode::SharedPtr & node,
```

Comment:
I think the node should be given in the constructor. If we want to initialize multiple times, ideally we're not playing with nodes.